Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder.py

```diff
@@ -0,0 +1,614 @@
+# ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import fnmatch
+import itertools
+import json
+import re
+import sys
+import urllib
+from collections import OrderedDict
+
+from webkitpy.layout_tests.controllers.test_result_writer import TestResultWriter
+from webkitpy.layout_tests.models.test import Test
+from webkitpy.w3c.common import TEMPLATED_TEST_HEADER
+
+supported_test_extensions = (
+    ".htm",
+    ".html",
+    ".mht",
+    ".php",
+    ".pl",
+    ".py",
+    ".shtml",
+    ".svg",
+    ".xht",
+    ".xhtml",
+    ".xml",
+)
+assert len(supported_test_extensions) == len(set(supported_test_extensions))
+assert list(supported_test_extensions) == sorted(supported_test_extensions)
+
+supported_reference_extensions = (
+    ".htm",
+    ".html",
+    ".svg",
+    ".xht",
+    ".xhtml",
+    ".xml",
+)
+assert len(supported_reference_extensions) == len(set(supported_reference_extensions))
+assert list(supported_reference_extensions) == sorted(supported_reference_extensions)
+assert set(supported_reference_extensions) < set(supported_test_extensions)
+
+
+skipped_directories = {
+    ".svn",
+    "_svn",
+    "resources",
+    "support",
+    "script-tests",
+    "tools",
+    "reference",
+    "reftest",
+}
+
+# Files ending in -ref/-notref aren't considered expectations, but are skipped.
+skipped_test_suffixes = tuple(
+    "".join(pair)
+    for pair in itertools.product(
+        ("-expected", "-expected-mismatch", "-ref", "-notref"),
+        supported_reference_extensions,
+    )
+)
+
+digit_re = re.compile("([0-9]+)")
+
+
+def natsort(string_to_split):
+    split = digit_re.split(string_to_split)
+    # this converts the split numbers into tuples so that "01" < "1"
+    split[1::2] = [(int(i), i) for i in split[1::2]]
+    return split
+
+
+class LayoutTestFinder(object):
+    def __init__(self, fs, layout_tests_base_dir, baseline_search_paths):
+        """Find layout tests.
+
+        :param FileSystem fs: the current filesystem object
+        :param str layout_tests_base_dir: the base directory of LayoutTests
+            (see: Port.layout_tests_dir())
+        :param List[str] baseline_search_paths: the baseline search paths, from most to
+            least specific (see: Port.baseline_search_path(device_type))
+        """
+        self.fs = fs
+
+        layout_tests_base_dir = fs.normpath(fs.realpath(layout_tests_base_dir))
+        baseline_search_paths = [
+            fs.normpath(fs.realpath(bsp)) for bsp in baseline_search_paths
+        ]
+
+        self.layout_tests_base_dir = layout_tests_base_dir
+        self.baseline_search_paths = baseline_search_paths
+
+        # XXX: both this and the realpath calls above mean __init__ reads the
+        # filesystem, and object data reflects FS state this time; are we okay with
+        # this?
```

Comment:
This seems reasonable, I struggle to think how we would structure this except to capture filesystem state on invocation.